### PR TITLE
Seba/full re design

### PR DESCRIPTION
**Pull request by argos83**
Ok, there're still some improvements to be done in terms of documentation, unit tests, and decisions to be made or revisited regarding status flows (e.g. shall a test status be skipped if all iterations were skipped?). But these changes reflect now the concept of Contexts which help with controlling and traceability.
@estemendoza let's have a skype quick sync so I can guide you through the changes.
Once this gets merged, the next thing to implement is a runner, i.e. the logic of taking cmd line arguments, finding test cases and data dynamically (e.g. via tags filters), build a suite on the fly with those tests and executing it.

### CODE REVIEW

File: marvin/core/test_runner.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    def _execute(self):
+        self._do_block('setup', self._data.setup_data(), TestSetupStartedEvent, TestSetupEndedEvent)
+
+        for it_data in self._data.iteration_data():
+            self._do_block('run', it_data, TestIterationStartedEvent, TestIterationEndedEvent)
+
+        self._do_block('tear_down', self._data.tear_down_data(), TestTearDownStartedEvent, TestTearDownEndedEvent)
+
+    def _do_block(self, block_type, data, started_event_class, ended_event_class):
+        if self._should_skip_block(block_type):
```
Comment (estemendoza) [inline marvin/core/test_runner.py]: Can you explain what this line does?

File: marvin/core/context.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    def __init__(self, parent_context=None):
+        self._ctx = parent_context
+        self._publisher = self.ctx.publisher if self.ctx else Publisher()
+        self._sub_context_results = {
+            Status.PASS: 0,
+            Status.FAIL: 0,
+            Status.SKIP: 0
+        }
+
```
Comment (estemendoza) [inline marvin/core/context.py]: can we use `_sub_context_results` property as base to create the summary report at the end of an execution?

File: marvin/core/test_runner.py (hunk 1 above)
Comment (argos83) [inline marvin/core/test_runner.py]: The setup phase, each of the iterations, are the tear down phase are blocks. This line is to decide whether the block should be executed or not. The current rules (which we can discuss and change if necessary) are:
  * The setup block is always executed (i.e. this call will always return False for `block_type='setup'`).
  * Iteration blocks are not executed if the setup phase resulted in FAIL, or SKIP.
  * The destroy block is not executed only if the setup phase resulted in SKIP (I'm not completely if this makes sense though)

File: marvin/core/context.py (hunk 2 above)
Comment (argos83) [inline marvin/core/context.py]: Sure. I can wrap that in a `@property` decorated method. Can you think of a better name than `sub_context_results`?

Comment (estemendoza) [inline marvin/core/context.py]: Well, even it is wrapped in a context, it all belongs to an "execution" right? So maybe something like `execution_status` should be fine. What do you think?

File: marvin/core/test_runner.py (hunk 1 above)
Comment (estemendoza) [inline marvin/core/test_runner.py]: Thanks, it makes sense for now, we can change it later if we need it.

File: marvin/core/context.py (hunk 2 above)
Comment (argos83) [inline marvin/core/context.py]: I called it `context_summary` since the context status may be anything regardless the subcontext results

**Pull request merged**